### PR TITLE
fix validate_no_duplicates so that dict and ndarray won't cause errors

### PR DESCRIPTION
**Pull request by wenlongzhao094**
## Summary by CodeRabbit

* **Bug Fixes**
  * Duplicate detection now reliably handles complex/unhashable cell values (0‑D and multi‑D arrays, nested dicts/lists, sets/frozensets, custom iterables and other edge types), preventing errors and improving duplicate reporting.

* **New Features**
  * Added robust safe dataset concatenation with built‑in validation for empty/None inputs and schema/context mismatches.

* **Refactor**
  * Duplicate‑check preprocessing consolidated into a single global normalization step.

* **Reliability**
  * Validation produces clearer, more robust failure behavior for varied datasets.

* **Tests**
  * Expanded coverage across many edge cases and fallback paths.

### CODE REVIEW

File: src/sdg_hub/core/utils/datautils.py [hunk 1]
```diff
@@ -39,27 +40,20 @@ def validate_no_duplicates(dataset: Dataset) -> None:
...
-                        )
-                    )
-            duplicate_count = int(df.duplicated(keep="first").sum())
+    def make_hashable(x):
+        if isinstance(x, dict):
+            return tuple(sorted((k, make_hashable(v)) for k, v in x.items()))
+        elif isinstance(x, (list, tuple, set)):
+            return tuple(make_hashable(i) for i in x)
+        elif isinstance(x, np.ndarray):
+            return tuple(make_hashable(i) for i in x.tolist())
```
Comment (mihirathale98) [inline src/sdg_hub/core/utils/datautils.py]: Hey @wenlongzhao094,

can you add another condition from the previous code which handles objs which are iterable, basically the hasattr(c, "__iter__") condition , will help handle cases which are of custom types in case users have that in their custom data

Comment (wenlongzhao094) [reply src/sdg_hub/core/utils/datautils.py]: Sure

Comment (wenlongzhao094) [reply src/sdg_hub/core/utils/datautils.py]: Fixed in new commit: fa654aa

File: src/sdg_hub/core/utils/datautils.py [hunk 2]
```diff
@@ -39,27 +40,27 @@ def validate_no_duplicates(dataset: Dataset) -> None:
...
+            return x
+        elif isinstance(x, dict):
+            return tuple(sorted((k, make_hashable(v)) for k, v in x.items()))
+        elif hasattr(x, "__iter__"):
+            # list, tuple, set, ndarray, or other customized iterables
+            return tuple(make_hashable(i) for i in x)
+
+    # Apply to the whole dataframe to ensure every cell is hashable
+    df = df.map(make_hashable)
+
```
[the remark this answers is not in the document]
Comment (wenlongzhao094) [reply src/sdg_hub/core/utils/datautils.py]: applymap is depreciated. The latest package has map.

File: src/sdg_hub/core/utils/datautils.py [hunk 3]
```diff
@@ -39,27 +40,27 @@ def validate_no_duplicates(dataset: Dataset) -> None:
...
+    def make_hashable(x):
+        if is_hashable(x):
+            # int, float, str, bytes, None etc. are already hashable
+            return x
+        elif isinstance(x, dict):
+            return tuple(sorted((k, make_hashable(v)) for k, v in x.items()))
+        elif hasattr(x, "__iter__"):
+            # list, tuple, set, ndarray, or other customized iterables
+            return tuple(make_hashable(i) for i in x)
+
```
[the remark this answers is not in the document]
Comment (wenlongzhao094) [reply src/sdg_hub/core/utils/datautils.py]: Incorporated.

File: src/sdg_hub/core/utils/datautils.py (hunk 1 above)
Comment (wenlongzhao094) [reply src/sdg_hub/core/utils/datautils.py]: Addressed. Latest commit: e8ddde0.

**Comment by eshwarprasadS** [comment]
@mihirathale98 can you confirm if this issue / PR is already covered by this closed PR? https://github.com/Red-Hat-AI-Innovation-Team/sdg_hub/pull/399

**Comment by mihirathale98** [comment]
@coderabbitai review

**Comment by wenlongzhao094** [comment]
> @mihirathale98 can you confirm if this issue / PR is already covered by this closed PR? #399

@eshwarprasadS Dictionaries and ndarrays both throw errors in the #399 fix. Both have to be fixed as in this PR. Otherwise the main function of LLMChat breaks: the prompt messages prepared by the PromptBuilderBlock are ndarray of dictionaries and cause errors in LLMChatBlock's validation.

**Review by shivchander — APPROVED** [review]
